File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        if(!head){
            return false;
        }

        let slow = head;
        let fast = head.next;

        while(fast && fast.next){
            if(fast.val == slow.val){
                return true;
            }

            fast = fast.next.next;
            slow = slow.next;
        }

        return false;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        i = 0
        curr = head
        while curr:
            i += 1
            curr = curr.next

        curr = head
        prev = None
        dummy = ListNode(None, head)
        
        for x in range(i - n + 1):
            if x == i - n:
                temp = curr.next
                curr.next = None
                if prev:
                    prev.next = temp
                else:
                    dummy.next = temp
            else:
                prev = curr
                curr = curr.next

        return dummy.next






